lensless/trainable_recon.py

```diff
@@ -0,0 +1,143 @@
+# #############################################################################
+# trainable_recon.py
+# =================
+# Authors :
+# Yohann PERRON [yohann.perron@gmail.com]
+# #############################################################################
+
+from lensless.recon import ReconstructionAlgorithm
+
+try:
+    import torch
+
+except ImportError:
+    raise ImportError("Pytorch is require to use trainable reconstruction algorithms.")
+
+
+class TrainableReconstructionAlgorithm(ReconstructionAlgorithm, torch.nn.Module):
+    """
+    Abstract class for defining lensless imaging reconstruction algorithms with trainable parameters.
+
+    The following abstract methods need to be defined:
+
+    * ``_update``: updating state variables at each iterations.
+    * ``reset``: reset state variables.
+    * ``_form_image``: any pre-processing that needs to be done in order to view the image estimate, e.g. reshaping or clipping.
+
+    One advantage of deriving from this abstract class is that functionality for
+    iterating, saving, and visualization is already implemented, namely in the
+    ``apply`` method.
+
+    Consequently, using a reconstruction algorithm that derives from it boils down
+    to four steps:
+
+    1. Creating an instance of the reconstruction algorithm.
+    2. Training the algorithm
+    3. Setting the data.
+    4. Applying the algorithm.
+
+
+    """
+
+    def __init__(self, psf, dtype=None, n_iter=1, **kwargs):
+        """
+        Base constructor. Derived constructor may define new state variables
+        here and also reset them in `reset`.
+
+        Parameters
+        ----------
+
+            psf : :py:class:`~torch.Tensor`
+                Point spread function (PSF) that models forward propagation.
+                2D (grayscale) or 3D (RGB) data can be provided and the shape will
+                be used to determine which reconstruction (and allocate the
+                appropriate memory).
+            dtype : float32 or float64
+                Data type to use for optimization.
+            n_iter : int
+                Number of iterations for unrolled algorithm.
+        """
+        assert isinstance(psf, torch.Tensor)
+        self.is_torch = True
+
+        self.n_iter = n_iter
+        super(TrainableReconstructionAlgorithm, self).__init__(psf, dtype=dtype, n_iter=1, **kwargs)
+
+    def batch_call(self, batch):
+        """
+        Method for performing iterative reconstruction on a batch of images.
+        This implementation simply calls `apply` on each image in the batch.
+        Training algorithms are expected to override this method with a properly vectorized implementation.
```

Comment:
This serves as a compatibility layer. It is however questionable whether or not it must be kept as the performance penalty of using this (instead of a proper implementation) is quite big (up to 5 times slower)

Reply:
remove implementation and leave as abstract